README.md
# Pomo

## Prerequisites

- `Node > v18.12.1`
- `Yarn > v1.22.9`

## Getting Started

`yarn start` will build the app and use `webpack-dev-server` to hot reload
any changes you make to watched files.

## Project Structure

- `dist/` this is final distribution folder that is meant to be manually or
automatically deployed to a web server or host.
- `public/` this contains the files which need to be included in the distribution
but do not need to be compiled or altered at build time.
- `src/` this contains the files that should be compiled or altered at build time

### Automagic to be aware of

- The webpack HTML plugin automatically compiles `.ejs` so the `src/index.ejs` is
used implicitly.


CHANGELOG.md
## v1.0.0-alpha004

### Features

- Plays an alarm sound when a cycle is completed [change](https://github.com/ominestre/pomo/pull/18)
  - The alarm sound was created using Reaper and Spire Synth

### Bugs

- Fixed an issue where the timer interval was setup incorrectly resulting in the state being fixed at the point of component render. Uses React's `useEffect` to re-evaluate the conditions of the interval whenever the subscribed state changes. [change](https://github.com/ominestre/pomo/pull/18)


### Build and Tools

- Configures Webpack and Typescript to support importing audio [change](https://github.com/ominestre/pomo/pull/18)
- Fixed the build script executing subshell git commands as a result of a poorly formatted string [change](https://github.com/ominestre/pomo/pull/12)

## v1.0.0-alpha003

### Build and Tools

- Fixes tag deployment workflow to use `main` branch instead of trying to find a branch with same name as tag [change](https://github.com/ominestre/pomo/pull/11/commits/f3e634aa4734949f50fde50f3e7870496cc1912a)
- Moves docs folder out of `src/` since it's not being built into a doc-site [change](https://github.com/ominestre/pomo/pull/11)
- Updates build script to set `NODE_ENV` to `production` for webpack

## v1.0.0-alpha002

### Tasks UI

- Adds maximum width to task columns [change](https://github.com/ominestre/pomo/pull/10/commits/4977f438d18ce496074be83151b25df040abff63)
- Adds color to task edit buttons [change](https://github.com/ominestre/pomo/pull/10/commits/35522edb2f0b80b08f3eb7069d84f79ab28bf40d)
- Changes the card cursor to indicate when drag and drop is enabled [change](https://github.com/ominestre/pomo/pull/10/commits/5c816fdf6d72049f47758e10d2021cc7c4abe9c0)
- Changes card inputs to be multi-line textareas instead of inputs [change](https://github.com/ominestre/pomo/pull/10/commits/8472f5da3126df5d36bfaa29fbd1825cd2f91906)

### Build and Tools

- Fixes sourcemaps only showing TS compiled scripts instead of original source [change](https://github.com/ominestre/pomo/pull/10/commits/e9693befd85c4ce6b2226f1d177dfedce4f046da)

## v1.0.0-alpha001

Finally at a point of deploying something that can actually be hosted and used so it's time to start
tracking changed! Yay!

### Features

- Creates Timer component
  - Has state management to handle swapping between work and break modes
  - Has UI to demonstrate when a timer is active or paused
  - Has controls for manually swapping timer modes, pausing, playing, and resetting
- Creates Kanban task board component
  - Can create and edit new cards
  - Can drag cards to other columns or other positions in the board (drop zones highlighted on drag start)
  - KNOWN BUG: Google Chrome's clickable zone for dragging a card isn't the entire card
- Creates Pager component
  - Only displays the Kanban board now but this will handle swapping what is visible in the main content area


src/features/timer/timer.test.ts
import {
  describe,
  expect,
  test,
} from '@jest/globals';

import timerReducer, {
  TimerState,
  initialState,
  resetTimer,
  stopTimer,
  startTimer,
  tick,
  completeCycle,
} from './timer.state';

describe('Timer state interactions', () => {
  describe('Timer mode switching', () => {
    test('Should swap from work mode to short break', () => {
      expect(timerReducer(initialState, completeCycle()))
        .toMatchObject({
          timerMode: 'shortBreak',
          cyclesCompleted: 1,
          intervalID: null,
          isTimerActive: false,
          minutes: initialState.sessionLength.shortBreak,
          seconds: 0,
        });
    });

    test('Should swap from short break to work mode', () => {
      const state: TimerState = { ...initialState, cyclesCompleted: 1, timerMode: 'shortBreak' };

      expect(timerReducer(state, completeCycle()))
        .toMatchObject({
          timerMode: 'work',
          cyclesCompleted: 1,
          intervalID: null,
          isTimerActive: false,
          minutes: initialState.sessionLength.work,
          seconds: 0,
        });
    });

    test('Should swap from work mode to long break', () => {
      const state: TimerState = { ...initialState, cyclesCompleted: 3, timerMode: 'work' };

      expect(timerReducer(state, completeCycle()))
        .toMatchObject({
          timerMode: 'longBreak',
          cyclesCompleted: 4,
          intervalID: null,
          isTimerActive: false,
          minutes: initialState.sessionLength.longBreak,
          seconds: 0,
        });
    });

    test('Should swap from long break to work', () => {
      const state: TimerState = { ...initialState, cyclesCompleted: 4, timerMode: 'longBreak' };

      expect(timerReducer(state, completeCycle()))
        .toMatchObject({
          timerMode: 'work',
          cyclesCompleted: 4,
          intervalID: null,
          isTimerActive: false,
          minutes: initialState.sessionLength.work,
          seconds: 0,
        });
    });
  });

  describe('Timer ticking', () => {
    test('Should tick down 1 second and roll over a minute', () => {
      expect(timerReducer(initialState, tick()))
        .toMatchObject({ minutes: 24, seconds: 59 });
    });

    test('Should tick down 1 second', () => {
      const state: TimerState = { ...initialState, minutes: 22, seconds: 47 };

      expect(timerReducer(state, tick()))
        .toMatchObject({ minutes: 22, seconds: 46 });
    });

    test('Should do nothing if the timer is already completed', () => {
      const state: TimerState = { ...initialState, minutes: 0, seconds: 0 };

      expect(timerReducer(state, tick()))
        .toMatchObject({ minutes: 0, seconds: 0 });
    });
  });

  describe('Timer reset', () => {
    test('Should reset to 25 minutes in work mode', () => {
      const state: TimerState = { ...initialState, minutes: 23 };

      expect(timerReducer(state, resetTimer()))
        .toMatchObject({
          isTimerActive: false,
          minutes: 25,
          seconds: 0,
          intervalID: null,
        });
    });

    test('Should reset to 5 minutes in short break mode', () => {
      const state: TimerState = { ...initialState, minutes: 3, timerMode: 'shortBreak' };

      expect(timerReducer(state, resetTimer()))
        .toMatchObject({
          isTimerActive: false,
          minutes: 5,
          seconds: 0,
          intervalID: null,
        });
    });

    test('Should reset to 15 minutes in long break mode', () => {
      const state: TimerState = { ...initialState, minutes: 2, timerMode: 'longBreak' };

      expect(timerReducer(state, resetTimer()))
        .toMatchObject({
          isTimerActive: false,
          minutes: 15,
          seconds: 0,
          intervalID: null,
        });
    });
  });

  test('Should start and stop a new timer', () => {
    const state: TimerState = { ...initialState };
    const newState = timerReducer(state, startTimer());

    expect(newState).toMatchObject({ isTimerActive: true });

    expect(timerReducer(newState, stopTimer()))
      .toMatchObject({ isTimerActive: false, intervalID: null });
  });
});


.eslintrc.js
module.exports = {
  extends: 'airbnb-base',
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
    'react-hooks',
  ],
  env: {
    node: true,
    browser: true,
  },
  rules: {
    'arrow-parens': 'off',
    'no-multiple-empty-lines': ['error', {
      max: 2,
    }],
    'no-underscore-dangle': 'off',
    'import/no-unresolved': 'off',
    'import/extensions': 'off',
    'implicit-arrow-linebreak': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn',
  },
  overrides: [
    {
      files: ['src/features/**/*.state.ts'],
      rules: {
        /*  justification: Redux Toolkit uses Immer and the maintainers are fairly hostile
            about providing alternatives. It reduces accidental mutations but with it's
            WritableDraft it requires writing less functional code. Directly modifying input
            parameters is their recommended pattern. */
        'no-param-reassign': 'warn',
      },
    },
    {
      files: ['src/features/**/*.tsx'],
      rules: {
        /*  justification: It is a React pattern for conditional rendering to have an `&&` operator
            at the end of the line of the logic encapsulating an element */
        'operator-linebreak': 'off',
      },
    },
  ],
};


webpack.config.js
const { join: pathJoin } = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

const { NODE_ENV } = process.env;

module.exports = {
  mode: NODE_ENV || 'development',
  devtool: NODE_ENV === 'production' ? false : 'eval-cheap-module-source-map',
  devServer: {
    static: './dist',
  },
  resolve: {
    extensions: ['.js', '.ts', '.tsx'],
  },
  entry: './src/index.tsx',
  output: {
    path: NODE_ENV === 'production' ? pathJoin(__dirname, './dist/') : pathJoin(__dirname, './build/'),
    clean: true,
  },
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: [
                ['@babel/preset-env', { targets: 'defaults' }],
              ],
            },
          },
          'ts-loader',
        ],
        exclude: /node_modules/,
      },
      {
        test: /\.scss$/i,
        use: [
          MiniCssExtractPlugin.loader,
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
            },
          },
          'postcss-loader',
          'sass-loader',
        ],
      },
      {
        test: /\.css$/i,
        use: [
          MiniCssExtractPlugin.loader,
          {
            loader: 'css-loader',
            options: { importLoaders: 1 },
          },
          'postcss-loader',
        ],
      },
      {
        test: /\.(mp3)$/i,
        type: 'asset/resource',
      },
    ],
  },
  plugins: [
    new HtmlWebpackPlugin(),
    new MiniCssExtractPlugin(),
  ],
};


src/features/tasks/tasks.state.ts
import { createSlice } from '@reduxjs/toolkit';
import { v4 as uuidv4 } from 'uuid';

type ColumnName = 'todo' | 'doing' | 'done';

export interface Card {
  uuid: string,
  title: string,
  body: string,
  position: number,
  column: ColumnName,
  isEditMode: boolean,
}

interface AddCardPayload {
  payload: Pick<Card, 'title' | 'body' | 'column'>,
}

interface MoveCardPayload {
  payload: Pick<Card, 'uuid' | 'column' | 'position'>,
}

interface ChangeCardPayload {
  payload: {
    uuid: string,
    body?: string,
    title?: string,
  },
}

interface EditCardPayload {
  payload: {
    uuid: string,
    isEditMode: boolean,
  }
}

export type CardCollection = Record<string, Card>;

interface TasksState {
  cards: CardCollection,
}

const initialState: TasksState = {
  cards: {},
};

const reindexCards = (
  cards: CardCollection,
  newIndex: number,
  columnFilter: ColumnName,
): CardCollection => {
  Object.values(cards)
    .filter(card => card.column === columnFilter)
    .forEach(({ position, uuid }) => {
      if (position >= newIndex) cards[uuid].position = position + 1;
    });

  return cards;
};

export const tasksSlice = createSlice({
  name: 'tasks',

  initialState,

  reducers: {
    addCard: (state, { payload }: AddCardPayload) => {
      let { cards } = state;
      const uuid = uuidv4();

      cards = reindexCards(cards, 1, payload.column);

      cards[uuid] = {
        uuid,
        title: payload.title,
        body: payload.body,
        position: 1,
        column: payload.column,
        isEditMode: false,
      };
    },

    removeCard: (state, { payload }: { payload: string }) => {
      const cards = { ...state.cards };

      if (cards[payload]) delete cards[payload];

      return {
        ...state,
        cards,
      };
    },

    moveCard: (state, { payload }: MoveCardPayload) => {
      const { cards } = state;
      const indexedCards = reindexCards(cards, payload.position, payload.column);

      indexedCards[payload.uuid].column = payload.column;
      indexedCards[payload.uuid].position = payload.position;
    },

    changeCard: (state, { payload }: ChangeCardPayload) => {
      const target = state.cards[payload.uuid];
      const title = payload.title || target.title;
      const body = payload.body || target.body;

      state.cards[payload.uuid] = {
        ...target,
        title,
        body,
        isEditMode: false,
      };
    },

    editCard: (state, { payload }: EditCardPayload) => {
      state.cards[payload.uuid].isEditMode = payload.isEditMode;
    },
  },
});

export const {
  addCard,
  removeCard,
  moveCard,
  changeCard,
  editCard,
} = tasksSlice.actions;

export default tasksSlice.reducer;


src/features/timer/timer.state.ts
import { createSlice } from '@reduxjs/toolkit';

type TimerMode = 'work' | 'shortBreak' | 'longBreak';

interface SessionLengths {
  work: number,
  shortBreak: number,
  longBreak: number,
}

export interface TimerState {
  minutes: number,
  seconds: number,
  isTimerActive: boolean,
  isTimerPaused: boolean,
  timerMode: TimerMode,
  cyclesCompleted: number,
  sessionLength: SessionLengths,
}

export const initialState: TimerState = {
  minutes: 25,
  seconds: 0,
  isTimerActive: false,
  isTimerPaused: false,
  timerMode: 'work',
  cyclesCompleted: 0,
  sessionLength: {
    work: 25,
    shortBreak: 5,
    longBreak: 15,
  },
};

export const timerSlice = createSlice({
  name: 'timer',

  initialState,

  reducers: {
    completeCycle: state => {
      let newTimerMode: TimerMode = 'work';
      const { timerMode, cyclesCompleted } = state;

      if (timerMode === 'work') {
        if (cyclesCompleted !== 0 && (cyclesCompleted + 1) % 4 === 0) {
          newTimerMode = 'longBreak';
        } else {
          newTimerMode = 'shortBreak';
        }
      }

      return {
        ...state,
        cyclesCompleted: state.timerMode === 'work'
          ? state.cyclesCompleted + 1
          : state.cyclesCompleted,
        isTimerActive: false,
        seconds: 0,
        minutes: state.sessionLength[newTimerMode],
        timerMode: newTimerMode,
      };
    },

    setCycleMode: (state, { payload }: { payload: TimerMode}) => ({
      ...state,
      timerMode: payload,
    }),

    tick: state => {
      const { minutes, seconds } = state;

      if (seconds === 0) {
        if (minutes === 0) return state;

        return { ...state, minutes: minutes - 1, seconds: 59 };
      }

      return { ...state, seconds: seconds - 1 };
    },

    startTimer: (state) => ({
      ...state,
      isTimerActive: true,
      isTimerPaused: false,
    }),

    stopTimer: state => ({
      ...state,
      isTimerActive: false,
      isTimerPaused: true,
      intervalID: null,
    }),

    resetTimer: (state: TimerState) => ({
      ...state,
      isTimerActive: false,
      isTimerPaused: false,
      minutes: state.sessionLength[state.timerMode],
      seconds: 0,
    }),
  },
});

export const {
  resetTimer,
  stopTimer,
  startTimer,
  tick,
  completeCycle,
  setCycleMode,
} = timerSlice.actions;

export default timerSlice.reducer;
